stackrox/acs-fleet-manager
# ROX-26276: update DNS after cluster reassignment

## Description

- Change FM to send `UPSERT` type requests to Route53, this way we can Update and Create CNAME records with the same operation
- Change Admin Endpoint for assign-cluster to reset route creation DB fields
- Change Admin Endpoint for assign-cluster to reset tenant to "provisioning" which will trigger resending of routes from FS -> Update in DB -> Upsert operation through cname_mgr -> Updated CNAME record after migration
- Added an additional DB field `EnteredProvisioning` to work around timeout issues, since CreatedAt is not a valid reference time for provisioning after reassignment. Adjusted preparing and provisioning operations to set and act according to the new field if present.
- Depends on #2056 and #2053 

## Checklist (Definition of Done)


## Test manual

- Adjusted integration tests to test DNS reset on AssignCentralCluster Admin API as well
- Tested on an infractl OSD as well:
  - Setup infra cluster
  - `make deploy/bootstrap deploy/dev`
  - Set `RHACS_CLUSTER_MIGRATION=true` on FM
  - Set `--enable-central-external-certificate=true`  on FM
  - Create a central
  - make sure DNS records are properly created
  - connect to the FM database to inspect routes_creation_id
  - Call the AssignCluster admin API endpoint (reassign to current cluster)
  - See in DB that all routes field are reseted and tenant is back to "provisioning" state
  - Verify tenant reaches ready state again, view that a new ID is stored in "routes_creation_id"
  - Verify that the change in AWS route53 has the correct timestamp (when new routes should be created)
  - Make sure to delete the tenant so that AWS resources are cleaned up
